tests/src/tracing/tracevalidation/inducedgc/InducedGC.cs

```diff
@@ -0,0 +1,92 @@
+using System;
+using System.IO;
+using Tracing.Tests.Common;
+using Microsoft.Diagnostics.Tracing;
+using Microsoft.Diagnostics.Tracing.Parsers;
+using Microsoft.Diagnostics.Tracing.Parsers.Clr;
+
+namespace Tracing.Tests
+{
+    public static class TraceValidationInducedGC
+    {
+        private static int InducedGCIterations = 10;
+
+        public static int Main(string[] args)
+        {
+            bool pass = true;
+            bool keepOutput = false;
+
+            // Use the first arg as an output filename if there is one.
+            string outputFilename = null;
+            if (args.Length >= 1)
+            {
+                outputFilename = args[0];
+                keepOutput = true;
+            }
+            else
+            {
+                outputFilename = System.IO.Path.GetTempPath() + Guid.NewGuid().ToString() + ".netperf";
+            }
+
+            try
+            {
+                Console.WriteLine("\tStart: Enable tracing.");
+                TraceControl.EnableDefault(outputFilename);
+                Console.WriteLine("\tEnd: Enable tracing.\n");
+
+                Console.WriteLine("\tStart: Generate some events.");
+                for(int i=0; i<InducedGCIterations; i++)
+                {
+                    GC.Collect();
+                }
+                Console.WriteLine("\tEnd: Generate some events.\n");
+
+                Console.WriteLine("\tStart: Disable tracing.");
+                TraceControl.Disable();
+                Console.WriteLine("\tEnd: Disable tracing.\n");
+
+                Console.WriteLine("\tStart: Process the trace file.");
+                int matchingEventCount = 0;
+                int nonMatchingEventCount = 0;
+                using (var trace = TraceEventDispatcher.GetDispatcherFromFileName(outputFilename))
+                {
+                    string gcReasonInduced = GCReason.Induced.ToString();
+                    string providerName = "Microsoft-Windows-DotNETRuntime";
+                    string gcTriggeredEventName = "GC/Triggered";
+
+                    trace.Clr.GCTriggered += delegate(GCTriggeredTraceData data)
+                    {
+                        if(gcReasonInduced.Equals(data.Reason.ToString()) &&
+                           providerName.Equals(data.ProviderName) &&
+                           gcTriggeredEventName.Equals(data.EventName))
+                        {
```

Comment:
I wonder if in case of this condition there could be some problems with `providerName` and `gcTriggeredEventName`? I mean that there are induced GC events, but with wrong metadata? Perhaps we could revert the order of asserts to:

```cs
Assert.Equal(nameof(nonMatchingEventCount), nonMatchingEventCount, 0);
Assert.Equal(nameof(matchingEventCount), InducedGCIterations, matchingEventCount);
```

Reply:
I'm not sure I follow here.  It is possible that a bug could arise that would cause the metadata to be incorrect, but it would still be caught here.  The advantage that you could get out of switching these is that you'd be able to easily see if a non-induced GC occurred even though the test doesn't expect one.  But, I think it both cases you're going to have to collect a trace to figure out what changed.  Let me know if I've mis-understood.

Reply:
As far I understood there is an environment, where we induce GC 10 times, but only one event is fired.

We test CoreCLR and assume that TraceEvent library 2.0.2 has no bugs. 

My point here is: what if CoreCLR is ok, but there are some bugs in TraceEvent?

I would just relax the condition from:

```cs
if(gcReasonInduced.Equals(data.Reason.ToString()) &&
                           providerName.Equals(data.ProviderName) &&
                           gcTriggeredEventName.Equals(data.EventName))
                        {
```

to simple `if(gcReasonInduced.Equals(data.Reason.ToString()))`

and change the order of asserts to:

```cs
Assert.Equal(nameof(nonMatchingEventCount), nonMatchingEventCount, 0);
Assert.Equal(nameof(matchingEventCount), InducedGCIterations, matchingEventCount);
```

because if the first assert `Assert.Equal(nameof(matchingEventCount), InducedGCIterations, matchingEventCount);` fails, we get an exception and stop on first error. So we don;t know what was the value of `nonMatchingEventCount` in this mysterious case. (1 induced gc event for 10x GC.Collect)

Reply:
Ah, gotcha.  The previous test (eventpipetrace) actually had asserts in place and incremented the event count regardless of the payload.  So, I suspect that if that bug still exists, it will show up no matter what the condition looks like and nonMatchingEventCount will still be zero.  One thing that might be worth doing though is printing the values before asserting or setting up the test so that it only passes if the values match.  I like the assert mechanism but I also like to be able to get more data printed to the console so we know what happened without a re-run.

Reply:
Let's see what these runs come back with - I'm open to modifying the tests to help flush out errors now and in the future.